File: 백준/2번문제/이채현/Q11286.py
# 맨 처음 생각은 모든 값에 절대값 취하고 그 값을 ordering 하는 걸 생각함
# 그런데 어찌됐든 원래 값을 뱉어내야 하고, -1과 1이 있으면 더 작은 -1?을 뱉어야 하므로, 튜플이나 딕셔너리를 생각함.
# 그래서 얘도 딕셔너리처럼 생각함.
# 예를 들어, dict = {'-1':1, 1:1, 2:2, 3:3} 뭐 이런식으로 접근하려 함.
# 그런데 자꾸 에러 남..!!
# 그래서 찾아보니까 딕셔너리가 아니라 heapq이라는 구조가 있고, 이걸 이용하면 해결 가능
# heapq은 (절대값, 원래값) 이렇게 넣으면 절대값 기준으로 정렬 가능
# 

import sys
import heapq

n = int(input())
q = []

for i in range(n):
    a = int(sys.stdin.readline().rstrip())
    if a != 0:
        heapq.heappush(q, (abs(a), a))
    else:
        if not q:
            print(0)
        else:
            print(heapq.heappop(q)[1])

File: 백준/2번문제/최원준/Q11286.java
// https://www.acmicpc.net/problem/
package 백준;

/*


1. 아이디어 :
    우선순위큐(힙)을 사용. [절대값, 기존 값] 형태로 힙에 저장하고 꺼냅니다.
2. 시간복잡도 :
O( )
3. 자료구조 :
    우선순위큐(힙)

 */
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.PriorityQueue;

class Q11286 {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static String input;
    static String[] inputs;
    static int[] ninputs;

    public static void main(String[] args) throws IOException {
        PriorityQueue<int[]> pq = new PriorityQueue<>((a,b) -> {
            if (a[0] == b[0]) return a[1]-b[1];
            return a[0]-b[0];
        }); // abs, ori

        int n = Integer.parseInt(br.readLine());
        for (int i=0; i<n; i++) {
            int num = Integer.parseInt(br.readLine());
            if (num == 0) {
                if (pq.isEmpty()) {
                    System.out.println(0);
                } else {
                    System.out.println(pq.poll()[1]);
                }
            } else {
                pq.offer(new int[]{Math.abs(num), num});
            }
        }

    }
}

/*
import sys
import heapq
#sys.setrecursionlimit(1000000)
# input = sys.stdin.readline
input = lambda: sys.stdin.readline().rstrip()

def solution(n):
    min_heap = []
    for i in range(n):
        num = int(input())
        if num == 0:
            if min_heap:
                print(heapq.heappop(min_heap)[1])
            else:
                print(0)
        else:
            heapq.heappush(min_heap, [abs(num), num])

if __name__ == '__main__':
    solution(int(input().rstrip()))

 */